src/hotspot/share/utilities/bitMap.hpp

```diff
@@ -385,6 +394,155 @@ class BitMap {
 #endif
 };
 
+// Implementation support for bitmap iteration.
+struct BitMap::IteratorImpl {
+  const BitMap* _map;
+  idx_t _cur_beg;
+  idx_t _cur_end;
+
+  void assert_not_empty() const NOT_DEBUG_RETURN;
+
+  // Constructs an empty iterator.
+  IteratorImpl();
+
+  // Constructs an iterator for map, over the range [beg, end).
+  IteratorImpl(const BitMap* map, idx_t beg, idx_t end);
+
+  // Returns true if the remaining iteration range is empty.
+  bool is_empty() const;
+
+  // Returns the index of the first set bit in the remaining iteration range.
+  // precondition: !is_empty()
+  idx_t first() const;
+
+  // Returns the index of the last set bit in the remaining iteration range.
+  // precondition: !is_empty()
+  idx_t last() const;
```

Comment:
The comments for `first()` and `last()` are a little misleading. The invariant that the `first()`/`last()` is a set bit is only true for `first()` in the forward iterator and for `last()` in the reverse iterator. 

It seems more like a contract that if you use `first()` then you must have created `IteratorImpl(const BitMap* map, idx_t beg, idx_t end)` with `beg` being a set bit. And if you use `last()` then you must have created it with `end` being a set bit.

Reply:
Yes, that's correct.  Some of the comments for IteratorImpl didn't get properly updated.  I'll clean that up.